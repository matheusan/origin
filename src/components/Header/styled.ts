import styled from 'styled-components';

export const SingleTitle: JSX.Element = styled.h1`
  width: 100%;
  margin: ${props => (props.noMargin ? '0' : '48px 0 24px 0')};
  padding: 0;

  text-align: center;
  font-weight: 400;
  font-size: ${props => (props.size ? `${props.size}px` : '2em')};
  font-family: ${props => props.theme.fonts.primary};
  color: ${props =>
    props.hard
      ? props.theme.colors.blueGray900
      : props.theme.colors.brandColorPrimary};

  @media only screen and (max-width: 380px) {
    margin: 32px 0 24px 0;
    font-size: 18px;
  }
`;

export const IconWrapper: JSX.Element = styled.header`
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: flex-start;
  align-items: center;
  align-content: stretch;
  margin-bottom: 28px;
`;

export const IconTitle: JSX.Element = styled.h1`
  width: 100%;
  margin: 0 0 0 16px;
  padding: 0;

  line-height: 29px;
  font-size: 24px;
  font-weight: 500;
  font-family: ${props => props.theme.fonts.secondary};
  color: ${props => props.theme.colors.blueGray900};
`;

export const IconDescription: JSX.Element = styled.h2`
  width: 100%;
  margin: 0 0 0 16px;
  padding: 0;

  line-height: 24px;
  font-size: 16px;
  font-weight: 400;
  font-family: ${props => props.theme.fonts.primary};
  color: ${props => props.theme.colors.blueGray400};
`;
