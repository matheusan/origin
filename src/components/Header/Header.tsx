import * as React from 'react';

import * as S from './styled';
import DisplayIcon, { Icons } from '../Icon';
import findByType from '../../utils/findByType';

const Title = () => null;
const Description = () => null;

export interface HeaderProps {
  icon?: Icons;
  hard?: boolean;
}

const Header: React.SFC<HeaderProps> = ({
  icon = null,
  hard = false,
  children
}): JSX.Element => {
  const title = findByType(children, Title);
  const description = findByType(children, Description);
  if (icon === null && !title && !description) {
    return <S.SingleTitle hard={hard}>{children}</S.SingleTitle>;
  }

  if (icon !== null && children) {
    return (
      <S.SingleTitle hard={hard} noMargin>
        <DisplayIcon name={icon} />
        <br />
        {children}
      </S.SingleTitle>
    );
  }

  return (
    <S.IconWrapper>
      {icon !== null && <DisplayIcon name={icon} />}
      <div>
        {title && <S.IconTitle>{title.props.children}</S.IconTitle>}
        {description && (
          <S.IconDescription>{description.props.children}</S.IconDescription>
        )}
      </div>
    </S.IconWrapper>
  );
};

Header.Title = Title;
Header.Description = Description;

export default Header;
