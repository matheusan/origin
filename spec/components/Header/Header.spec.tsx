import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import Header from '../../../src/components/Header';
import {
  SingleTitle,
  IconTitle,
  IconDescription
} from '../../../src/components/Header/styled';

describe('Header', () => {
  let wrapper: ShallowWrapper;

  it('renders', () => {
    wrapper = shallow(<Header>This is my header</Header>);
    expect(wrapper).toMatchSnapshot();
  });

  describe('when single title', () => {
    it('should render single title', () => {
      const text = 'This is my header';
      wrapper = shallow(<Header>{text}</Header>);
      const singleTitle = wrapper.find(SingleTitle);
      expect(singleTitle.length).toBe(1);
      expect(singleTitle.text()).toBe(text);
    });
  });

  describe('when icon title', () => {
    it('should render all elements', () => {
      wrapper = shallow(
        <Header icon="arrow">
          <Header.Title>My title</Header.Title>
          <Header.Description>Subdescription</Header.Description>
        </Header>
      );
      expect(wrapper).toMatchSnapshot();
    });

    it('should render title and description without icon', () => {
      wrapper = shallow(
        <Header>
          <Header.Title>My title</Header.Title>
          <Header.Description>Subdescription</Header.Description>
        </Header>
      );
      expect(wrapper.find('Icon').length).toBe(0);
    });

    it('should render only title', () => {
      wrapper = shallow(
        <Header>
          <Header.Title>My title</Header.Title>
        </Header>
      );
      expect(wrapper.find(IconTitle).length).toBe(1);
    });

    it('should render only description', () => {
      wrapper = shallow(
        <Header>
          <Header.Description>Description</Header.Description>
        </Header>
      );
      expect(wrapper.find(IconDescription).length).toBe(1);
    });
  });
});
