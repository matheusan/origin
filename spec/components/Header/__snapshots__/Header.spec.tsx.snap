// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders 1`] = `
<styled.h1
  hard={false}
>
  This is my header
</styled.h1>
`;

exports[`Header when icon title should render all elements 1`] = `
<styled.h1
  hard={false}
  noMargin={true}
>
  <Icon
    name="arrow"
  />
  <br />
  <Title>
    My title
  </Title>
  <Description>
    Subdescription
  </Description>
</styled.h1>
`;
